doc/rpc_marshalable_objects.md

```diff
@@ -237,8 +237,141 @@ These resources are released and the `IDisposable.Dispose()` method is invoked o
+1. Must not include any events.
+
+The proxy object created for the marshalable object will only implement the subset of interfaces described by `RpcMarshalableKnownSubType` attributes that are applied to the interface type used in the RPC method declaration. The receiver of the proxy can use the [is](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/is) operator to check if an optional interface is implemented. Casting a proxy object to an unimplemented interface will result in an `InvalidCastException`.
+
+### Use cases
+
+`RpcMarshalableKnownSubType` is useful in a few scenarios:
+1. adding new methods to an existing marshalable interface without breaking backward compatibility,
+1. creating an RPC method that return marshalable objects which implement different interfaces to accomodate for different behaviors,
+1. creating an RPC method that return marshalable objects which may optionally implement additional functionalities.
+
+For example, the following code shows how `RpcMarshalableKnownSubType` attributes can be added to the `ICounter` interface from the earlier sample.
+
+```cs
+[RpcMarshalable]
+[RpcMarshalableKnownSubType(typeof(IAdvancedCounter, subTypeCode: 1))]
+[RpcMarshalableKnownSubType(typeof(IDecrementable, subTypeCode: 2))]
+interface ICounter : IDisposable
+{
+    Task IncrementAsync(CancellationToken ct);
+
+    Task<int> GetCountAsync(CancellationToken ct);
+}
+
+[RpcMarshalable]
+interface IAdvancedCounter : ICounter
+{
+    Task IncrementAsync(int count, CancellationToken ct);
+}
+
+[RpcMarshalable]
+interface IDecrementable : IDisposable
+{
+    Task DecrementAsync(CancellationToken ct);
+}
+```
+
+This change doesn't require any modification to the declarations of RPC methods that exchange `ICounter` object and is fully backward compatible. A client that uses an earlier RPC contract definition will simply ignore any unknow additional interface.
+
+A marshalable object implementing `ICounter` can also implement `IAdvancedCounter`, `IDecrementable`, both or neither.
+
+### Caveats
+
+#### Backward compatibility
+
+The `subTypeCode` values used in `RpcMarshalableKnownSubType` attributes are used as part of the wire protocol. While it is backward compatible to remove an `RpcMarshalableKnownSubType` attribute from an interface, its `subTypeCode` value should never be reused to add a different additional interface to the same interface declaration.
+
+#### Method name conflicts and non-marshalable interfaces
+
+When a method is invoked on a marshalable object proxy, it is traslated into an RPC call to the best-matching interface method.
+
+For example, given the interfaces defined below:
+- a call to `((IFoo)proxy).DoFooAsync()` would result in an RPC call to the `DoFooAsync` method as defined by the `IFoo` interface.
+- a call to `((IBaz)proxy).DoFooAsync()` would result in an RPC call to the `DoFooAsync` method as defined by the `IBaz` interface, if the marshalable object implements `IBaz`.
```

Comment:
What happens otherwise? I assume proxy wouldn't even implement IBaz if the underlying object didn't right?

Reply:
Exactly right. That's explained above on line 251:
> The proxy object created for the marshalable object will only implement the subset of interfaces described by `RpcMarshalableKnownSubType` attributes that are applied to the interface type used in the RPC method declaration. The receiver of the proxy can use the `is` operator to check if an optional interface is implemented.
And 286
> A client that uses an earlier RPC contract definition will simply ignore any unknow additional interface.